File: algo/src/main/java/com/hartwig/actin/algo/evaluation/general/IsFemale.java
package com.hartwig.actin.algo.evaluation.general;

import com.hartwig.actin.PatientRecord;
import com.hartwig.actin.algo.datamodel.Evaluation;
import com.hartwig.actin.algo.datamodel.EvaluationResult;
import com.hartwig.actin.algo.datamodel.ImmutableEvaluation;
import com.hartwig.actin.algo.evaluation.EvaluationFactory;
import com.hartwig.actin.algo.evaluation.EvaluationFunction;
import com.hartwig.actin.clinical.datamodel.Gender;

import org.jetbrains.annotations.NotNull;

public class IsFemale implements EvaluationFunction {

    IsFemale() {
    }

    @NotNull
    @Override
    public Evaluation evaluate(@NotNull PatientRecord record) {
        EvaluationResult result = record.clinical().patient().gender() == Gender.FEMALE ? EvaluationResult.PASS : EvaluationResult.FAIL;

        ImmutableEvaluation.Builder builder = EvaluationFactory.unrecoverable().result(result);
        if (result == EvaluationResult.FAIL) {
            builder.addFailSpecificMessages("Patient is not female");
            builder.addFailGeneralMessages("Inadequate gender");
        } else if (result == EvaluationResult.PASS) {
            builder.addPassSpecificMessages("Patient is female");
            builder.addPassGeneralMessages("Adequate gender");
        }

        return builder.build();
    }
}

File: algo/src/main/kotlin/com/hartwig/actin/algo/evaluation/general/IsFemale.kt
package com.hartwig.actin.algo.evaluation.general

import com.hartwig.actin.algo.evaluation.EvaluationFactory
import com.hartwig.actin.algo.evaluation.EvaluationFunction
import com.hartwig.actin.datamodel.PatientRecord
import com.hartwig.actin.datamodel.algo.Evaluation
import com.hartwig.actin.datamodel.clinical.Gender

class IsFemale : EvaluationFunction {

    override fun evaluate(record: PatientRecord): Evaluation {
        return when (record.patient.gender) {
            Gender.FEMALE -> EvaluationFactory.pass("Gender is female")
            Gender.MALE -> EvaluationFactory.fail("Gender is not female")
            null -> EvaluationFactory.undetermined("Undetermined if gender is female")
        }
    }
}